File: crypto-samples/crypto_samples.py
"""
Sample cryptographic code for testing the scanner (Python).
Run: cryptoscan scan ./crypto-samples to see detection results.

This file demonstrates various cryptographic patterns in Python.
"""

from cryptography.hazmat.primitives import hashes, serialization
from cryptography.hazmat.primitives.asymmetric import rsa, ec, ed25519, padding
from cryptography.hazmat.primitives.ciphers import Cipher, algorithms, modes
from cryptography.hazmat.backends import default_backend
import hashlib
import os


# ============================================================================
# QUANTUM VULNERABLE - Asymmetric Cryptography
# ============================================================================

def generate_rsa_key(key_size: int = 2048):
    """Generate RSA key pair - QUANTUM VULNERABLE"""
    private_key = rsa.generate_private_key(
        public_exponent=65537,
        key_size=key_size,
        backend=default_backend()
    )
    return private_key


def generate_weak_rsa_key():
    """Generate weak RSA-1024 key - CRITICAL (classically weak)"""
    return rsa.generate_private_key(
        public_exponent=65537,
        key_size=1024,  # Too small!
        backend=default_backend()
    )


def generate_ecdsa_key():
    """Generate ECDSA key - QUANTUM VULNERABLE"""
    return ec.generate_private_key(ec.SECP256R1(), default_backend())


def generate_ed25519_key():
    """Generate Ed25519 key - QUANTUM VULNERABLE"""
    return ed25519.Ed25519PrivateKey.generate()


def sign_with_ed25519(private_key, message: bytes) -> bytes:
    """Sign with Ed25519 - QUANTUM VULNERABLE"""
    return private_key.sign(message)


# ============================================================================
# QUANTUM PARTIAL - Symmetric Cryptography
# ============================================================================

def encrypt_aes_256_gcm(key: bytes, plaintext: bytes) -> bytes:
    """Encrypt with AES-256-GCM - QUANTUM PARTIAL (acceptable)"""
    iv = os.urandom(12)
    cipher = Cipher(algorithms.AES(key), modes.GCM(iv), backend=default_backend())
    encryptor = cipher.encryptor()
    ciphertext = encryptor.update(plaintext) + encryptor.finalize()
    return iv + ciphertext + encryptor.tag


def encrypt_aes_128(key: bytes, plaintext: bytes) -> bytes:
    """Encrypt with AES-128 - needs upgrade to AES-256"""
    iv = os.urandom(16)
    cipher = Cipher(algorithms.AES(key[:16]), modes.CBC(iv), backend=default_backend())
    encryptor = cipher.encryptor()
    return iv + encryptor.update(plaintext) + encryptor.finalize()


# ============================================================================
# BROKEN/WEAK - Should be replaced immediately
# ============================================================================

def hash_md5(data: bytes) -> bytes:
    """Hash with MD5 - BROKEN (collision attacks exist)"""
    return hashlib.md5(data).digest()


def hash_sha1(data: bytes) -> bytes:
    """Hash with SHA-1 - WEAK (collision attacks demonstrated)"""
    return hashlib.sha1(data).digest()


def encrypt_des(key: bytes, plaintext: bytes) -> bytes:
    """Encrypt with DES - BROKEN (56-bit key)"""
    iv = os.urandom(8)
    cipher = Cipher(algorithms.TripleDES(key), modes.CBC(iv), backend=default_backend())
    encryptor = cipher.encryptor()
    return iv + encryptor.update(plaintext) + encryptor.finalize()


# ============================================================================
# QUANTUM SAFE - Hash functions
# ============================================================================

def hash_sha256(data: bytes) -> bytes:
    """Hash with SHA-256 - QUANTUM SAFE"""
    return hashlib.sha256(data).digest()


def hash_sha512(data: bytes) -> bytes:
    """Hash with SHA-512 - QUANTUM SAFE"""
    return hashlib.sha512(data).digest()


# ============================================================================
# PATTERNS THAT SHOULD BE FILTERED
# ============================================================================

def log_crypto_info():
    """Logs crypto info - should be filtered (print statements)"""
    print("Using Ed25519 for authentication")
    print("Encrypting with AES-256-GCM")


def get_crypto_metadata() -> dict:
    """Returns metadata - should be filtered (string labels)"""
    return {
        "algorithm": "ed25519",
        "key_type": "asymmetric",
        "auth_method": "ecdsa",
    }


def validate_key(key_type: str) -> None:
    """Validates key - should be filtered (error messages)"""
    if not key_type:
        raise ValueError("key_type must be a valid Ed25519 or RSA identifier")

File: crypto-samples/CryptoSamples.java
package cryptosamples;

import javax.crypto.Cipher;
import javax.crypto.KeyGenerator;
import javax.crypto.SecretKey;
import javax.crypto.spec.GCMParameterSpec;
import java.security.*;
import java.security.spec.ECGenParameterSpec;

/**
 * Sample cryptographic code for testing the scanner (Java).
 * Run: cryptoscan scan ./crypto-samples to see detection results.
 */
public class CryptoSamples {

    // ========================================================================
    // QUANTUM VULNERABLE - Asymmetric Cryptography
    // ========================================================================

    /**
     * Generate RSA key pair - QUANTUM VULNERABLE
     * Remediation: Migrate to ML-KEM (FIPS 203)
     */
    public static KeyPair generateRSAKeyPair() throws Exception {
        KeyPairGenerator keyGen = KeyPairGenerator.getInstance("RSA");
        keyGen.initialize(2048);
        return keyGen.generateKeyPair();
    }

    /**
     * Generate weak RSA-1024 key - CRITICAL (classically weak)
     */
    public static KeyPair generateWeakRSAKey() throws Exception {
        KeyPairGenerator keyGen = KeyPairGenerator.getInstance("RSA");
        keyGen.initialize(1024); // Too small!
        return keyGen.generateKeyPair();
    }

    /**
     * Generate ECDSA key - QUANTUM VULNERABLE
     * Remediation: Migrate to ML-DSA (FIPS 204)
     */
    public static KeyPair generateECDSAKeyPair() throws Exception {
        KeyPairGenerator keyGen = KeyPairGenerator.getInstance("EC");
        keyGen.initialize(new ECGenParameterSpec("secp256r1"));
        return keyGen.generateKeyPair();
    }

    /**
     * Sign with ECDSA - QUANTUM VULNERABLE
     */
    public static byte[] signWithECDSA(PrivateKey privateKey, byte[] data) throws Exception {
        Signature signature = Signature.getInstance("SHA256withECDSA");
        signature.initSign(privateKey);
        signature.update(data);
        return signature.sign();
    }

    // ========================================================================
    // QUANTUM PARTIAL - Symmetric Cryptography
    // ========================================================================

    /**
     * Encrypt with AES-256-GCM - QUANTUM PARTIAL (acceptable)
     */
    public static byte[] encryptAES256GCM(byte[] key, byte[] plaintext) throws Exception {
        SecretKey secretKey = new javax.crypto.spec.SecretKeySpec(key, "AES");
        Cipher cipher = Cipher.getInstance("AES/GCM/NoPadding");
        byte[] iv = new byte[12];
        SecureRandom random = new SecureRandom();
        random.nextBytes(iv);
        cipher.init(Cipher.ENCRYPT_MODE, secretKey, new GCMParameterSpec(128, iv));
        return cipher.doFinal(plaintext);
    }

    /**
     * Encrypt with AES-128 - needs upgrade to AES-256
     */
    public static SecretKey generateAES128Key() throws Exception {
        KeyGenerator keyGen = KeyGenerator.getInstance("AES");
        keyGen.init(128); // Should be 256
        return keyGen.generateKey();
    }

    // ========================================================================
    // BROKEN/WEAK - Should be replaced immediately
    // ========================================================================

    /**
     * Hash with MD5 - BROKEN (collision attacks exist)
     */
    public static byte[] hashMD5(byte[] data) throws Exception {
        MessageDigest md = MessageDigest.getInstance("MD5");
        return md.digest(data);
    }

    /**
     * Hash with SHA-1 - WEAK (collision attacks demonstrated)
     */
    public static byte[] hashSHA1(byte[] data) throws Exception {
        MessageDigest md = MessageDigest.getInstance("SHA-1");
        return md.digest(data);
    }

    /**
     * Encrypt with DES - BROKEN (56-bit key)
     */
    public static byte[] encryptDES(byte[] key, byte[] plaintext) throws Exception {
        Cipher cipher = Cipher.getInstance("DES/CBC/PKCS5Padding");
        // DES is completely broken
        return cipher.doFinal(plaintext);
    }

    /**
     * Encrypt with 3DES - WEAK (deprecated)
     */
    public static byte[] encrypt3DES(byte[] key, byte[] plaintext) throws Exception {
        Cipher cipher = Cipher.getInstance("DESede/CBC/PKCS5Padding");
        return cipher.doFinal(plaintext);
    }

    // ========================================================================
    // QUANTUM SAFE - Hash functions
    // ========================================================================

    /**
     * Hash with SHA-256 - QUANTUM SAFE
     */
    public static byte[] hashSHA256(byte[] data) throws Exception {
        MessageDigest md = MessageDigest.getInstance("SHA-256");
        return md.digest(data);
    }

    /**
     * Hash with SHA-512 - QUANTUM SAFE
     */
    public static byte[] hashSHA512(byte[] data) throws Exception {
        MessageDigest md = MessageDigest.getInstance("SHA-512");
        return md.digest(data);
    }

    // ========================================================================
    // PATTERNS THAT SHOULD BE FILTERED
    // ========================================================================

    /**
     * Logs crypto info - should be filtered (print statements)
     */
    public static void logCryptoInfo() {
        System.out.println("Using RSA-2048 for key exchange");
        System.out.println("Signing with ECDSA P-256");
    }

    /**
     * Validates input - should be filtered (error messages)
     */
    public static void validateKeyType(String keyType) {
        if (keyType == null) {
            throw new IllegalArgumentException("keyType must be a valid RSA or ECDSA key");
        }
    }
}
